src/java.base/share/classes/jdk/internal/classfile/CodeBuilder.java

```diff
@@ -0,0 +1,1386 @@
+
+    /**
+     * Add an "if-then-else" block that is conditional on the value(s) on top of the operand stack
+     * in accordance with the given opcode.
+     * <p>
+     * The {@link BlockCodeBuilder#breakLabel()} for each block corresponds to the
+     * end of the "else" block.
+     *
+     * @param opcode the operation code for a branch instructions that accepts one or two operands on the stack
+     * @param thenHandler handler that receives a {@linkplain BlockCodeBuilder} to
+     *                    generate the body of the {@code if}
+     * @param elseHandler handler that receives a {@linkplain BlockCodeBuilder} to
+     *                    generate the body of the {@code else}
+     * @return this builder
+     * @throws java.lang.IllegalArgumentException if the operation code is not for a branch instruction that accepts
+     * one or two operands
+     */
+    default CodeBuilder ifThenElse(Opcode opcode,
+                                   Consumer<BlockCodeBuilder> thenHandler,
+                                   Consumer<BlockCodeBuilder> elseHandler) {
+        if (opcode.kind() != Opcode.Kind.BRANCH || opcode.primaryTypeKind() == TypeKind.VoidType) {
+            throw new IllegalArgumentException("Illegal branch opcode: " + opcode);
+        }
+
+        Label breakLabel = newLabel();
+        BlockCodeBuilderImpl thenBlock = new BlockCodeBuilderImpl(this, breakLabel);
+        BlockCodeBuilderImpl elseBlock = new BlockCodeBuilderImpl(this, breakLabel);
+        branchInstruction(BytecodeHelpers.reverseBranchOpcode(opcode), elseBlock.startLabel());
+        thenBlock.start();
+        thenHandler.accept(thenBlock);
+        if (thenBlock.reachable())
+            thenBlock.branchInstruction(Opcode.GOTO, thenBlock.breakLabel());
+        thenBlock.end();
+        elseBlock.start();
+        elseHandler.accept(elseBlock);
+        elseBlock.end();
+        labelBinding(breakLabel);
+        return this;
+    }
+
+    /**
+     * A builder to add catch blocks.
+     *
+     * @see #trying
+     */
+    sealed interface CatchBuilder permits CatchBuilderImpl {
+        /**
+         * Adds a catch block that catches an exception of the given type.
+         * <p>
+         * The caught exception will be on top of the operand stack when the catch block is entered.
+         * <p>
+         * If the type of exception is {@code null} then the catch block catches all exceptions.
+         *
+         * @param exceptionType the type of exception to catch.
+         * @param catchHandler handler that receives a {@linkplain CodeBuilder} to
+         *                     generate the body of the catch block.
+         * @return this builder
+         * @throws java.lang.IllegalArgumentException if an existing catch block catches an exception of the given type.
+         * @see #catchingMulti
+         * @see #catchingAll
+         */
+        CatchBuilder catching(ClassDesc exceptionType, Consumer<BlockCodeBuilder> catchHandler);
```

Comment:
I imagine there are name clashes with Java keyword, hence the `ing` in the names. That said, this should probably revisited in a later bikeshedding round - as in the current form, the code builder API has most method names that are nouns (the thing being built) but it also has some verbs in there (trying, catching) which seem odd.

Reply:
Please raise the naming convention discussion at classfile-api-dev at openjdk.org
Thanks.